### PR TITLE
Add Inception-ResNet v2 to application

### PR DESCRIPTION
**Pull request by myutwo150**
Adds the Inception-ResNet v2 architecture to the application module, with weights extracted and converted from the [TF-slim implementation](https://github.com/tensorflow/models/tree/master/slim#pre-trained-models).

Brief summary:
- Add `keras.applications.inception_resnet_v2.InceptionResNetV2()`
- Update test for applications
- Update docs for applications

Possible issues:

1. Model weights: right now the .h5 weight files are in my own public repo.
Once the model is included into Keras, I think it's better to move the weight files here.
Is there any suggested way to do it?

2. Layer naming:
    Layer naming follows TF-slim because it's easier to debug when converting the weights.
    But it's not that critical once the code and weight files are included into Keras.
    Right now the solution is not so elegant, and makes the code look a bit cluttered.
    Is there a better way to do it, or should I remove the layer names?

3. The `dropout_keep_rate` argument:
    The original paper and TF-slim use keep rate, but Keras uses drop rate (`= 1 - dropout_keep_rate`) in the `Dropout` layer. Should I change the API to use drop rate?

### CODE REVIEW

**Comment by myutwo150** [comment]
CNTK tests failed because there seems to be a bug in `cntk_backend.bias_add()` that breaks the dynamic axes.

```python
x = Input(shape=(None, None, 3))
conv_no_bias = Conv2D(32, 1, use_bias=False)(x)
conv_with_bias = Conv2D(32, 1, use_bias=True)(x)
print(x.shape)  # (-1, -1, 3)
print(conv_no_bias.shape)  # (-1, -1, 32)
print(conv_with_bias.shape)  # (-1, 1, 32)
```

Even with a fixed input size, the predictions are very wrong with CNTK backend. Theano and Tensorflow both work fine. I've tested and `InceptionV3` has the same prediction issue under CNTK backend.

I think I'll just skip the test and raise an error if this model is used with CNTK until someone more familiar with it fixes the backend issue(s?).

**Comment by fchollet** [comment]
This is useful, we'll merge it.

> I think I'll just skip the test and raise an error if this model is used with CNTK until someone more familiar with it fixes the backend issue(s?).

That is fine. @souptc, are you aware of the issue?

> Model weights: right now the .h5 weight files are in my own public repo.

I'll do it after we merge this.

> layer names

For the sake of readability, avoid name-generating functions. Use in-place string formatting. Avoid naming every layer, instead only give name to some key layers (e.g. block outputs). Take a look at InceptionV3 for an example.

> The dropout_keep_rate argument:

Do not include regularizers. These models are primarily for inference. Other applications don't have regularizers.

**Comment by souptc** [comment]
The convolution with dynamic axis issue is on our list and we properly fix it soon in Sept. But I need to take a look at the wrong prediction issue with InceptionV3. We do plan to unblock cntk keras with more keras applications, please go ahead to temporary disable cntk with this application and we will try our best to make cntk work with it soon.

File: keras/applications/inception_resnet_v2.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        bn_axis = 1 if K.image_data_format() == 'channels_first' else 3
+        bn_name = None if name is None else name + '_bn'
+        x = BatchNormalization(axis=bn_axis, scale=False, name=bn_name)(x)
+    if activation is not None:
+        ac_name = None if name is None else name + '_ac'
+        x = Activation(activation, name=ac_name)(x)
+    return x
+
+
+def _inception_resnet_block(x, scale, block_type, block_idx, activation='relu'):
```
Comment (fchollet) [inline keras/applications/inception_resnet_v2.py]: No need to make this a private function (e.g. `conv2d_bn` isn't).

File: keras/applications/inception_resnet_v2.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    in the paper, controlled by the `block_type` argument (which is the
+    block name used in the official TF-slim implementation):
+        - Inception-ResNet-A: `block_type='block35'`
+        - Inception-ResNet-B: `block_type='block17'`
+        - Inception-ResNet-C: `block_type='block8'`
+
+    # Arguments
+        x: input tensor.
+        scale: scaling factor to scale the residuals before adding
+            them to the shortcut branch.
```
Comment (fchollet) [inline keras/applications/inception_resnet_v2.py]: Add more info, e.g. explicit computation of the residual output

File: keras/applications/inception_resnet_v2.py [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+        block_type: `'block35'`, `'block17'` or `'block8'`, determines
+            the network structure in the residual branch.
+        block_idx: used for generating layer names.
```
Comment (fchollet) [inline keras/applications/inception_resnet_v2.py]: Not descriptive enough

File: keras/applications/inception_resnet_v2.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+        activation: name of the activation function to use at the end
```
Comment (fchollet) [inline keras/applications/inception_resnet_v2.py]: This could be any function, not just a name

File: docs/templates/applications.md [hunk 5]
```diff
@@ -459,6 +461,64 @@ These weights are released under [the Apache License](https://github.com/tensorf
 
 -----
 
+## InceptionResNetV2
+
+
+```python
+keras.applications.inception_resnet_v2.InceptionResNetV2(include_top=True, weights='imagenet', input_tensor=None, input_shape=None, pooling=None, classes=1000, dropout_keep_prob=0.8)
```
Comment (fchollet) [inline docs/templates/applications.md]: The docs still seems to refer to `dropout_keep_prob`

File: keras/applications/inception_resnet_v2.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+                      pooling=None,
+                      classes=1000):
+    """Instantiates the Inception-ResNet v2 architecture.
+
+    Optionally loads weights pre-trained on ImageNet.
+    Note that when using TensorFlow, for best performance you should
+    set `"image_data_format": "channels_last"` in your Keras config
+    at `~/.keras/keras.json`.
+
+    The model and the weights are compatible with both TensorFlow and Theano
```
Comment (Dref360) [inline keras/applications/inception_resnet_v2.py]: Sure about theano? I only see tf at line 370.

Comment (myutwo150) [reply keras/applications/inception_resnet_v2.py]: I've tested the model on sample images and the result seems reasonable. I think weight conversion from TF to TH is handled in `model.load_weights()`, since backend information is included in the weight files.

Comment (fchollet) [reply keras/applications/inception_resnet_v2.py]: Correct, in theory it should work with both. What results do you see?

Comment (myutwo150) [reply keras/applications/inception_resnet_v2.py]: For example, given a random truck image, the prediction is:

With TensorFlow:

    [[('n03417042', 'garbage_truck', 0.77079743),
      ('n03126707', 'crane', 0.05554872),
      ('n03796401', 'moving_van', 0.051750395),
      ('n04252225', 'snowplow', 0.028239522),
      ('n04467665', 'trailer_truck', 0.01623963)]]

With Theano:

    [[('n03417042', 'garbage_truck', 0.76859063),
      ('n03126707', 'crane', 0.059462097),
      ('n03796401', 'moving_van', 0.054951463),
      ('n04252225', 'snowplow', 0.027199779),
      ('n04467665', 'trailer_truck', 0.015053505)]]

The probabilities are not exactly the same, but the order is correct (also tested on `InceptionV3` and `ResNet50`, and observed similar results). Changing the `image_data_format` does not affect the probabilities at all.

Comment (fchollet) [reply keras/applications/inception_resnet_v2.py]: The discrepancy is likely due to a known issue with how Theano and TF use different implementations for unpadded pooling operations. It's fine.

File: docs/templates/applications.md [hunk 7]
```diff
@@ -189,17 +191,17 @@ and a top-5 validation accuracy of 0.945.
 
 Note that this model is only available for the TensorFlow backend,
 due to its reliance on `SeparableConvolution` layers. Additionally it only supports
-the data format "channels_last" (height, width, channels).
+the data format `channels_last` (height, width, channels).
```
Comment (fchollet) [inline docs/templates/applications.md]: This is string, not a code keyword (such as a variable name). Please format as a string. Generally avoid any changes that aren't directly related to your PR.

File: docs/templates/applications.md [hunk 8]
```diff
@@ -244,17 +246,17 @@ keras.applications.vgg16.VGG16(include_top=True, weights='imagenet', input_tenso
 VGG16 model, with weights pre-trained on ImageNet.
 
 This model is available for both the Theano and TensorFlow backend, and can be built both
-with "channels_first" data format (channels, height, width) or "channels_last" data format (height, width, channels).
+with `channels_first` data format (channels, height, width) or `channels_last` data format (height, width, channels).
```
Comment (fchollet) [inline docs/templates/applications.md]: Same: there are strings

File: docs/templates/applications.md [hunk 9]
```diff
@@ -300,17 +302,17 @@ keras.applications.vgg19.VGG19(include_top=True, weights='imagenet', input_tenso
 VGG19 model, with weights pre-trained on ImageNet.
 
 This model is available for both the Theano and TensorFlow backend, and can be built both
-with "channels_first" data format (channels, height, width) or "channels_last" data format (height, width, channels).
+with `channels_first` data format (channels, height, width) or `channels_last` data format (height, width, channels).
```
Comment (fchollet) [inline docs/templates/applications.md]: Same

File: docs/templates/applications.md [hunk 10]
```diff
@@ -357,18 +359,18 @@ keras.applications.resnet50.ResNet50(include_top=True, weights='imagenet', input
 ResNet50 model, with weights pre-trained on ImageNet.
 
 This model is available for both the Theano and TensorFlow backend, and can be built both
-with "channels_first" data format (channels, height, width) or "channels_last" data format (height, width, channels).
+with `channels_first` data format (channels, height, width) or `channels_last` data format (height, width, channels).
```
Comment (fchollet) [inline docs/templates/applications.md]: Same

File: docs/templates/applications.md [hunk 11]
```diff
@@ -413,18 +415,18 @@ keras.applications.inception_v3.InceptionV3(include_top=True, weights='imagenet'
 Inception V3 model, with weights pre-trained on ImageNet.
 
 This model is available for both the Theano and TensorFlow backend, and can be built both
-with "channels_first" data format (channels, height, width) or "channels_last" data format (height, width, channels).
+with `channels_first` data format (channels, height, width) or `channels_last` data format (height, width, channels).
```
Comment (fchollet) [inline docs/templates/applications.md]: Same

File: docs/templates/applications.md [hunk 12]
```diff
@@ -459,6 +461,62 @@ These weights are released under [the Apache License](https://github.com/tensorf
...
+
+
+```python
+keras.applications.inception_resnet_v2.InceptionResNetV2(include_top=True, weights='imagenet', input_tensor=None, input_shape=None, pooling=None, classes=1000)
+```
+
+Inception-ResNet V2 model, with weights pre-trained on ImageNet.
+
+This model is available for both the Theano and TensorFlow backend (but not CNTK), and can be built both
+with `channels_first` data format (channels, height, width) or `channels_last` data format (height, width, channels).
```
Comment (fchollet) [inline docs/templates/applications.md]: Same

File: tests/keras/applications/applications_test.py [hunk 13]
```diff
@@ -235,6 +235,43 @@ def test_inceptionv3_variable_input_channels():
...
+def test_inceptionresnetv2_pooling():
+    model = applications.InceptionResNetV2(weights=None, include_top=False, pooling='avg')
+    assert model.output_shape == (None, 1536)
+
+
+@keras_test
+@pytest.mark.skipif((K.backend() == 'cntk'),
+                    reason='InceptionResNetV2 is not supported on CNTK')
+def test_inceptionresnetv2_variable_input_channels():
+    input_shape = (1, None, None) if K.image_data_format() == 'channels_first' else (None, None, 1)
```
Comment (fchollet) [inline tests/keras/applications/applications_test.py]: This test is only every run with the default image data format (channels_last), and in fact you are hard-coding this expectation (with the shape `(None, None, None, 1536)`). Please add a test for `channels_first` (set and later restore the global image_data_format).

File: docs/templates/applications.md (hunk 7 above)
Comment (myutwo150) [reply docs/templates/applications.md]: Okay, I can change it back.

File: tests/keras/applications/applications_test.py (hunk 13 above)
Comment (myutwo150) [reply tests/keras/applications/applications_test.py]: Sure.

File: docs/templates/applications.md (hunk 12 above)
Comment (myutwo150) [reply docs/templates/applications.md]: Just to make sure, in current version (i.e. master branch) of this file, both "channels_first" (plain text) and \`channels_first\` (code keyword) are used. The code keyword format \`channels_first\` is used in the arguments section. Is it better to follow current usage, or unify them into one of "channels_first", \`channels_first\` or \`'channels_first'\`?

**Comment by abrad1212** [comment]
@fchollet Is this gonna be merged?

File: docs/templates/applications.md (hunk 12 above)
Comment (fchollet) [reply docs/templates/applications.md]: Let's go with `'channels_first'`, I guess, for maximum explicitness

**Comment by fchollet** [comment]
@abrad1212 sure we will merge it

**Comment by myutwo150** [comment]
Sync with master branch to resolve conflict on the model comparison table. Also added this model to the table.

**Comment by myutwo150** [comment]
Thanks a lot!

**Comment by vfdev-5** [comment]
I think it would be good to mention in the docs that presented architecture is not exactly the same as presented in the paper. See [this issue](https://github.com/tensorflow/models/issues/1553)
